### PR TITLE
[#118] Fix unsafe Json http api data instances

### LINKED ISSUE #118: Fix unsafe FomJSON/FromJSONKey/FromHttpApiData instances
# Clarification and motivation

Consider the `FieldName` data type:

```hs
newtype FieldName = UnsafeFieldName Text
  deriving stock (Show, Eq)
  deriving newtype (A.ToJSON, A.ToJSONKey, A.FromJSON, A.FromJSONKey, Hashable, Buildable, ToHttpApiData, FromHttpApiData)

allowedCharSet :: [Char]
allowedCharSet = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9'] ++ "-_;"

newFieldName :: Text -> Either BadFieldName FieldName
newFieldName t
  | T.null t =
      Left $ BadFieldName "Tags must contain at least 1 character"
  | T.any (`notElem` allowedCharSet) t =
      Left $ BadFieldName ("Tags can only contain the following characters: '" <> T.pack allowedCharSet <> "'")
  | otherwise = Right $ UnsafeFieldName t
```

This data type is a newtype wrapper around `Text` with the following invariants: the text is not empty and only contains the allowed characters. The constructor is prefixed with `Unsafe` to discourage its use - we should use the `newFieldName` smart constructor instead. The smart constructor enforces the invariants.

However, the newtype-derived `FromJSON` and `FromHttpApiData` instances don't use the smart constructor, they coerce through the `Unsafe` constructor.

```hs
> A.decode' @FieldName "\"\""
Just (UnsafeFieldName "")
```

We should add hand-written instances that use the smart constructor.
Let's do this for every newtype with a "smart constructor".

NOTE: the `newFieldName` error messages say "Tags must ..." instead of saying "Field names must ...", let's fix that too.

# Acceptance criteria

* The `FromJSON`, `FromJSONKey` and `FromHttpApiData` instances respect the types' invariants.
* The error messages in `newFieldName` do not mention tags

### PR DESCRIPTION
**Pull request by Kariel-Myrr**
### Problem:

`FieldName`, `EntryTag`, `PathSegment` used unsafe(default text) constructors
for `FromJSON`, `FromJSONKey`  and `FromHttpApiData`

### Solution:

Write new safe instances that use smart `newFieldName` constructor. 
Cover new functionality with bad-case tests
Also corrected previous Error msg.

## Related issue(s)

Fixed #118 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: tests/server-integration/Utils.hs [hunk 1]
```diff
@@ -159,45 +162,68 @@ makeCofferHeader =  BS.toStrict . encode $
...
+    Just "application/json;charset=utf-8" -> do
+      jsonBodyActual <- either assertFailure pure $ eitherDecodeStrict @Value bs
+      jsonBodyActual @?= jsonBodyExpected
+      responseStatus response @?= expectedStatus
+    Just contentType -> assertFailure $ "Not expected content-type heder : " <> (map (chr . fromEnum) $ BS.unpack contentType)
+    Nothing -> do
+      assertFailure "No content-type header were specified"
+
+unwarpStatusCodeError :: (Show res) => (Response () -> BS.ByteString -> IO()) -> Either Req.HttpException res -> IO()
+unwarpStatusCodeError validator = \case
```
Comment (dcastro) [inline tests/server-integration/Utils.hs:221]: "unwarp" -> "unwrap"

File: tests/server-integration/Tag/TagTest.hs [hunk 2]
```diff
@@ -9,6 +9,7 @@ module Tag.TagTest
   , unit_tag_tag_not_found
   , unit_tag_duplicate_tag
   , unit_tag_updates_only_entry's_modification_time
+  , addOrRemoveTag
```
Comment (dcastro) [inline tests/server-integration/Tag/TagTest.hs:12]: I don't think this export is needed

**Review by dcastro — COMMENTED** [review]
LGTM, only a couple of small comments!

File: tests/server-integration/Tag/TagTest.hs (hunk 2 above)
Comment (Kariel-Myrr) [reply tests/server-integration/Tag/TagTest.hs:12]: But I need it for Common.hs It's used to add incorrect tag

Comment (dcastro) [reply tests/server-integration/Tag/TagTest.hs:12]: Ah I see, in that case let's move it to the `Utils` module. Test suite should be independent, they shouldn't import from each other.

File: tests/server-integration/Utils.hs [hunk 3]
```diff
@@ -159,45 +162,68 @@ makeCofferHeader =  BS.toStrict . encode $
...
 
 processStatusCodeError :: (Show res) => Status -> Value -> Either Req.HttpException res -> IO ()
-processStatusCodeError expectedStatus jsonBodyExpected = \case
+processStatusCodeError expectedStatus jsonBodyExpected = unwarpStatusCodeError $ \response bs ->
+  case lookup "Content-Type" $ responseHeaders response of
+    Just "application/json;charset=utf-8" -> do
+      jsonBodyActual <- either assertFailure pure $ eitherDecodeStrict @Value bs
+      jsonBodyActual @?= jsonBodyExpected
+      responseStatus response @?= expectedStatus
+    Just contentType -> assertFailure $ "Not expected content-type heder : " <> (map (chr . fromEnum) $ BS.unpack contentType)
```
[the remark this answers is not in the document]
Comment (Kariel-Myrr) [reply tests/server-integration/Utils.hs]: done

File: tests/server-integration/Utils.hs [hunk 4]
```diff
@@ -159,45 +162,68 @@ makeCofferHeader =  BS.toStrict . encode $
...
+    Nothing -> do
+      assertFailure "No content-type header were specified"
```
[the remark this answers is not in the document]
Comment (Kariel-Myrr) [reply tests/server-integration/Utils.hs]: done

File: tests/server-integration/Utils.hs (hunk 1 above)
Comment (Kariel-Myrr) [reply tests/server-integration/Utils.hs]: done

File: tests/server-integration/Tag/TagTest.hs (hunk 2 above)
Comment (Kariel-Myrr) [reply tests/server-integration/Tag/TagTest.hs]: done

**Review by dcastro — APPROVED** [review]

**Pull request merged**